### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by blatinier**
Beginning of API auth module.
CR welcome :)
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394286%22%2C%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394578%22%2C%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394642%22%2C%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394702%22%2C%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101464255%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20e3e108fea0df3531dd550b2d3a7b2330d071d9c2%20back/api/auth.py%20158%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394642%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Implementation%20needed%22%2C%20%22created_at%22%3A%20%222017-02-15T22%3A03%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1086629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/blatinier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20back/api/auth.py%3AL1-265%22%7D%2C%20%22Pull%20e3e108fea0df3531dd550b2d3a7b2330d071d9c2%20back/api/auth.py%20264%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394702%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Implementation%20needed%22%2C%20%22created_at%22%3A%20%222017-02-15T22%3A03%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1086629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/blatinier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20back/api/auth.py%3AL1-265%22%7D%2C%20%22Pull%20e3e108fea0df3531dd550b2d3a7b2330d071d9c2%20back/models/user.py%2064%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101464255%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Maybe%20see%20to%20use%20an%20observer%20and%20do%20this%20in%20a%20pre_save%20method.%22%2C%20%22created_at%22%3A%20%222017-02-16T08%3A19%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1713404%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Far3t%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20back/models/user.py%3AL15-67%22%7D%2C%20%22Pull%20e3e108fea0df3531dd550b2d3a7b2330d071d9c2%20back/api/auth.py%20147%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394578%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Implementation%20needed%22%2C%20%22created_at%22%3A%20%222017-02-15T22%3A03%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1086629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/blatinier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20back/api/auth.py%3AL1-265%22%7D%2C%20%22Pull%20e3e108fea0df3531dd550b2d3a7b2330d071d9c2%20back/api/auth.py%2090%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/blatinier/onestpasdespigeons/pull/26%23discussion_r101394286%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Implementation%20needed%22%2C%20%22created_at%22%3A%20%222017-02-15T22%3A01%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1086629%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/blatinier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20back/api/auth.py%3AL1-265%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull e3e108fea0df3531dd550b2d3a7b2330d071d9c2 back/api/auth.py 90'></a> &nbsp;<b><a href='https://github.com/blatinier/onestpasdespigeons/pull/26#discussion_r101394286'>File: back/api/auth.py:L1-265</a></b>
- <a href='https://github.com/blatinier'></a> Implementation needed
- [ ] <a href='#crh-comment-Pull e3e108fea0df3531dd550b2d3a7b2330d071d9c2 back/api/auth.py 147'></a> &nbsp;<b><a href='https://github.com/blatinier/onestpasdespigeons/pull/26#discussion_r101394578'>File: back/api/auth.py:L1-265</a></b>
- <a href='https://github.com/blatinier'></a> Implementation needed
- [ ] <a href='#crh-comment-Pull e3e108fea0df3531dd550b2d3a7b2330d071d9c2 back/api/auth.py 158'></a> &nbsp;<b><a href='https://github.com/blatinier/onestpasdespigeons/pull/26#discussion_r101394642'>File: back/api/auth.py:L1-265</a></b>
- <a href='https://github.com/blatinier'></a> Implementation needed
- [ ] <a href='#crh-comment-Pull e3e108fea0df3531dd550b2d3a7b2330d071d9c2 back/api/auth.py 264'></a> &nbsp;<b><a href='https://github.com/blatinier/onestpasdespigeons/pull/26#discussion_r101394702'>File: back/api/auth.py:L1-265</a></b>
- <a href='https://github.com/blatinier'></a> Implementation needed
- [x] <a href='#crh-comment-Pull e3e108fea0df3531dd550b2d3a7b2330d071d9c2 back/models/user.py 64'></a> &nbsp;<b><a href='https://github.com/blatinier/onestpasdespigeons/pull/26#discussion_r101464255'>File: back/models/user.py:L15-67</a></b>
- <a href='https://github.com/Far3t'></a> Maybe see to use an observer and do this in a pre_save method.

<a href='https://www.codereviewhub.com/blatinier/onestpasdespigeons/pull/26?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/blatinier/onestpasdespigeons/pull/26?approve=1'></a>&nbsp;<a href='https://github.com/blatinier/onestpasdespigeons/pull/26'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by blatinier** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/blatinier/onestpasdespigeons/pull/26'>Refresh</a>.<a href='#crh-update-none'></a>

File: back/api/auth.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    This route will update an existing user's account settings.
+    The default accepted params are password and password_confirmation,
+    but this can be customized using the devise_parameter_sanitizer
+    system. If config.check_current_password_before_update is set
+    to :attributes the current_password param is checked before
+    any update, if it is set to :password the current_password
+    param is checked only if the request updates user password.
+    """
+    pass
+    # TODO
```
Comment (blatinier) [inline back/api/auth.py:90]: Implementation needed for account update

File: back/api/auth.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+
+@auth_bp.route("/<provider>", methods=["GET"])
+def sso(provider):
+    """
+    Set this route as the destination for client authentication.
+    Ideally this will happen in an external window or popup.
+    """
+    pass
+    # TODO
```
Comment (blatinier) [inline back/api/auth.py:147]: Implementation needed for SSO

File: back/api/auth.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+@auth_bp.route("/<provider>/callback", methods=["GET", "POST"])
+def sso_callback(provider):
+    """
+    Destination for the oauth2 provider's callback uri.
+    postMessage events containing the authenticated user's data
+    will be sent back to the main client window from this page.
+    """
+    pass
+    # TODO
```
Comment (blatinier) [inline back/api/auth.py:158]: Implementation needed for SSO callback

File: back/api/auth.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+def password_reset_confirm():
+    """
+    Verify user by password reset token.
+    This route is the destination URL for password reset confirmation.
+    This route must contain reset_password_token and redirect_url params.
+    These values will be set automatically by the confirmation email
+    that is generated by the password reset request.
+    """
+    pass
+    # TODO
```
Comment (blatinier) [inline back/api/auth.py:264]: Implementation needed for password reset callback

File: back/models/user.py [hunk 5]
```diff
@@ -14,4 +15,52 @@ class User(Document):
...
+        """
+        to_del = []
+        now = int(time.time())
+        for access_token, clients in self.tokens.items():
+            for client, expiry in clients.items():
+                if expiry < now:
+                    to_del.append((access_token, client))
+        for token in to_del:
+            self._invalidate_token(*token)
+        super().save()
```
Comment (Far3t) [inline back/models/user.py]: Maybe see to use an observer and do this in a pre_save method.

**Pull request closed without merging**